File: DS using python/linked-lists/doublyLL.py
class node:
    def __init__(
        self,
        emp_id: int = 0,
        emp_mobileno: int = 0,
        emp_sal: float = 0.0,
        emp_dept: str = "",
        nxt=None,
        prev=None,
    ):
        self.emp_id, self.emp_mobileno, self.emp_sal = emp_id, emp_mobileno, emp_sal
        self.emp_dept, self.nxt, self.prev = emp_dept, nxt, prev


class DLL:
    def __init__(self):
        self.head = self.last = None
        self.count = 0

    def create(self):
        emp_id = int(input("Enter employee id : "))
        if self.searchNode(emp_id) is not None:
            print("\nEmployee already exists in DLL\n")
            return None
        emp_mobileno, emp_sal, emp_dept = (
            int(input("Enter employee mobile no : ")),
            float(input("Enter employee salary: ")),
            input("Enter employee dept : "),
        )
        temp = node(emp_id, emp_mobileno, emp_sal, emp_dept)
        self.count += 1
        return temp

    def searchNode(self, emp_id):
        temp, ptr = None, self.head
        while ptr:
            if ptr.emp_id == emp_id:
                temp = ptr
                break
            ptr = ptr.nxt
        return temp

    def add(self):
        i = 1
        n = int(input("Enter number of employees you want to add: "))
        if self.head == None:
            self.last = self.head = self.create()
            i += 1
            print()
        while i <= n:
            print("---------------Enter Employee - %d details---------------\n" % i)
            self.insertAtEnd()
            i += 1
        print("Successfully updated the DLL\n")

    def displayDLL(self):
        temp = self.head
        print(
            "\n%-10s%-15s%-13s%s\n" % ("Emp id", "Emp_mobileno", "Emp sal", "Emp dept")
        )
        while temp:
            print(
                "%-10d%-15d%-13.2f%s"
                % (temp.emp_id, temp.emp_mobileno, temp.emp_sal, temp.emp_dept)
            )
            temp = temp.nxt
        print()

    def insertAtEnd(self):
        temp = self.create()
        if temp is None:
            return
        self.last.nxt = temp
        temp = self.last
        self.last = self.last.nxt
        self.last.prev = temp

    def deleteaAtEnd(self):
        if self.last.prev is not None:
            self.last = self.last.prev
            self.last.nxt = None
        else:
            self.head = self.last = None
        self.count -= 1
        print("\nDeleted successfully\n")

    def deleteAtFront(self):
        if self.head.nxt:
            self.head = self.head.nxt
            self.head.prev = None
        else:
            self.head = None
        self.count -= 1
        print("Deleted successfully\n")

    def insertAtFront(self):
        temp = self.create()
        if temp is None:
            return
        temp.nxt = self.head
        self.head.prev = temp
        self.head = temp
        print("Successfully updated the DLL\n")

    def insert(self):
        emp_id = int(input("Enter employee id after which you want to insert : "))
        temp = self.searchNode(emp_id)
        if temp is None:
            print("\nEmployee id entered doesn't exist in DLL\n")
        else:
            if temp.nxt != None:
                temp1 = self.create()
                if temp1 is None:
                    return
                temp1.nxt = temp.nxt
                temp1.prev = temp
                temp.nxt = temp1
                temp1.nxt.prev = temp1
            else:
                self.insertAtEnd()
            print("Successfully updated the DLL\n")

    def delet(self):
        emp_id = int(input("Enter employee id you want to delete : "))
        temp = self.searchNode(emp_id)
        if temp is None:
            print("Employee id entered doesn't exist in DLL\n")
        else:
            if temp.nxt is None:
                self.deleteaAtEnd()
            elif temp.prev is None:
                self.deleteAtFront()
            else:
                temp.prev.nxt = temp.nxt
                temp.nxt.prev = temp.prev
                self.count -= 1
            print("\nDeleted successfully\n")

    def todo(self, option):
        if option in range(1, 11):
            if option == 1:
                self.add()
            elif option == 2:
                print("Total members in DLL is %d" % self.count)
            elif option == 3:
                if self.count == 0:
                    print("There are no employees in DLL\n")
                else:
                    self.displayDLL()
            elif option == 4:
                if self.count == 0:
                    print("There are no employees in DLL\n")
                else:
                    self.insertAtFront()
            elif option == 5:
                if self.count == 0:
                    print("There are no employees in DLL\n")
                else:
                    self.insertAtEnd()
            elif option == 6:
                if self.count == 0:
                    print("There are no employees in DLL\n")
                else:
                    self.insert()
            elif option == 7:
                if self.count == 0:
                    print("There are no employees in DLL to delete\n")
                else:
                    self.deleteAtFront()
            elif option == 8:
                if self.count == 0:
                    print("There are no employees in DLL to delete\n")
                else:
                    self.deleteaAtEnd()
            elif option == 9:
                if self.count == 0:
                    print("There are no employees in DLL to delete\n")
                else:
                    self.delet()
            elif option == 10:
                print("\n------------------------Program ends------------------------")
        else:
            print("Enter a valid option !!\n")

    def operations(self):
        option = 0
        while option != 10:
            print(
                "\n1. Add members\n"
                + "2.  Count of nodes in DLL\n"
                + "3.  Display Employee data in DLL\n"
                "4.  Insert Employee at Front of the DLL\n"
                + "5.  Insert Employee at End of the DLL\n"
                + "6.  Insert Employee at a specific position in DLL\n"
                + "7.  Delete Employee at Front of the DLL\n"
                + "8.  Delete Employee at End of the DLL\n"
                + "9.  Delete Employee at a specific position in DLL\n"
                + "10. Exit\n"
            )
            option = int(input("\nEnter your option: "))
            self.todo(option)


if __name__ == "__main__":
    print("------------------------Program starts------------------------\n\n")
    l = DLL()
    l.operations()

File: DS using c++/linked-lists/doublyLL.cpp
#include <cstdio>
#include <iomanip>
#include <iostream>
#include <limits>
using namespace std;
const char l = '\n';
class Node {
   public:
    int emp_id, emp_mobileno;
    double emp_sal;
    string emp_dept;
    Node *next, *prev;
    Node(int emp_id = 0, int emp_mobileno = 0, double emp_sal = 0.0,
         string emp_dept = "", Node *next = NULL, Node *prev = NULL)
        : emp_id(emp_id),
          emp_mobileno(emp_mobileno),
          emp_sal(emp_sal),
          emp_dept(emp_dept),
          next(next),
          prev(prev) {}
};
class LinkedList {
   public:
    Node *head = NULL;
    Node *last = NULL;
    int count = 0;
    Node *searchNode(int id) {
        Node *temp = NULL;
        Node *ptr = head;
        while (ptr != NULL) {
            if (ptr->emp_id == id) {
                temp = ptr;
                break;
            }
            ptr = ptr->next;
        }
        return temp;
    }
    Node *create() {
        int id;
        printf("Enter employee id : ");
        cin >> id;
        if (searchNode(id) != NULL) {
            cout << l << "Employee already exists in DLL" << l;
            return NULL;
        }
        Node *temp = new Node();
        temp->emp_id = id;
        cout << "Enter employee mobile no, salary separated by space "
                ": ";
        cin >> temp->emp_mobileno >> temp->emp_sal;
        cin.ignore(numeric_limits<streamsize>::max(), '\n');
        cout << "Enter employee dept : ";
        cin >> temp->emp_dept;
        cout << l;
        count++;
        return temp;
    }
    void insertAtEnd() {
        Node *temp = create();
        if (temp == NULL) return;
        if (head->next == NULL) {
            last = temp;
            head->next = last;
            last->prev = head;
            return;
        }
        last->next = temp;
        temp = last;
        last = last->next;
        last->prev = temp;
    }
    void add() {
        int n, i = 1;
        cout << l << "Enter number of employees you want to add: ";
        cin >> n;
        Node *temp;
        if (head == NULL) {
            cout << "---------------Enter Employee - " << i
                 << " details---------------\n";
            temp = create();
            i++;
            head = temp;
            last = temp;
        }
        while (i <= n) {
            cout << "---------------Enter Employee - " << i
                 << " details---------------\n";
            insertAtEnd();
            i++;
        }
        cout << l << "Successfully updated the DLL" << l;
    }
    void displayDLL() {
        Node *temp = head;
        printf("\n%-10s%-15s%-13s%s\n", "Emp id", "Emp_mobileno",
               "Emp sal", "Emp dept");
        while (temp != NULL) {
            printf("%-10d%-15d%-13.2f", temp->emp_id,
                   temp->emp_mobileno, temp->emp_sal);
            cout << temp->emp_dept << l;
            temp = temp->next;
        }
        cout << l;
    }
    void del_main() {
        delete head;
        head = NULL;
        last = NULL;
    }
    void deleteaAtEnd() {
        if (last->prev != NULL) {
            last = last->prev;
            delete last->next;
            last->next = nullptr;
        } else
            del_main();
        count--;
        cout << l << "Deleted successfully" << l;
    }
    void deleteAtFront() {
        if (head->next != NULL) {
            head = head->next;
            delete head->prev;
            head->prev = nullptr;
        } else
            del_main();
        count--;
        cout << l << "Deleted successfully" << l;
    }
    void insertAtFront() {
        Node *temp = create();
        if (temp == NULL) return;
        head->prev = temp;
        temp = head;
        head = head->prev;
        head->next = temp;
        cout << "\nSuccessfully updated the DLL\n";
    }
    void insert() {
        int id;
        cout << "Enter employee id after which you want to insert : ";
        cin >> id;
        Node *temp = searchNode(id);
        if (temp == NULL) {
            cout << "\nEmployee id entered doesn't exist in DLL\n";
        } else {
            Node *temp1 = create();
            if (temp1 == NULL) return;
            if (temp->next != NULL) {
                temp1->next = temp->next;
                temp->next = temp1;
                temp1->prev = temp;
                temp1->next->prev = temp1;
            } else {
                temp->next = temp1;
                temp1->prev = temp;
            }
        }
        cout << l << "Successfully updated the DLL" << l;
    }
    void delet() {
        int emp_id;
        cout << "Enter employee id you want to delete : ";
        cin >> emp_id;
        Node *temp = searchNode(emp_id);
        if (temp == NULL) {
            cout << l << "Employee id entered doesn't exist in DLL"
                 << l;
        } else {
            if (temp->next == NULL) {
                deleteaAtEnd();
            } else if (temp->prev == NULL) {
                deleteAtFront();
            } else {
                temp->prev->next = temp->next;
                temp->next->prev = temp->prev;
                delete temp;
            }
            count--;
            cout << l << "Deleted successfully" << l;
        }
    }
    void finish() {
        Node *temp;
        while (head != NULL) {
            temp = head;
            head = head->next;
            delete temp;
        }
        head = nullptr;
        last = nullptr;
    }
    void todo(int option) {
        switch (option) {
            case 1:
                add();
                break;
            case 2:
                cout << l << "Total members in DLL is " << count << l;
                break;
            case 3:
                if (count == 0)
                    cout << l << "There are no employees in DLL" << l;
                else
                    displayDLL();
                break;
            case 4:
                if (count == 0)
                    cout << l << "There are no employees in DLL" << l;
                else
                    insertAtFront();
                break;
            case 5:
                if (count == 0)
                    cout << l << "There are no employees in DLL" << l;
                else
                    insertAtEnd();
                break;
            case 6:
                if (count == 0)
                    cout << l << "There are no employees in DLL" << l;
                else
                    insert();
                break;
            case 7:
                if (count == 0)
                    cout << l
                         << "There are no employees in DLL to delete"
                         << l;
                else
                    deleteAtFront();
                break;
            case 8:
                if (count == 0)
                    cout << l
                         << "There are no employees in DLL to delete"
                         << l;
                else
                    deleteaAtEnd();
                break;
            case 9:
                if (count == 0)
                    cout << l
                         << "There are no employees in DLL to delete"
                         << l;
                else
                    delet();
                break;
            case 10:
                finish();
                cout << l << l
                     << "------------------------Program "
                        "ends------------------------"
                     << l << l;
                break;
            default:
                cout << l << "Enter a valid option !!" << l;
        }
    }
    void operations() {
        int option;
        do {
            cout
                << ("\n1-> Add members\n"
                    "2->  Count of nodes in DLL\n"
                    "3->  Display Employee data in DLL\n"
                    "4->  Insert Employee at Front of the DLL\n"
                    "5->  Insert Employee at End of the DLL\n"
                    "6->  Insert Employee at a specific position in "
                    "DLL\n"
                    "7->  Delete Employee at Front of the DLL\n"
                    "8->  Delete Employee at End of the DLL\n"
                    "9->  Delete Employee at a specific position in "
                    "DLL\n"
                    "10-> Exit\n");
            cout << ("\nEnter your option: ");
            cin >> option;
            fflush(stdin);
            todo(option);
        } while (option != 10);
    }
};
int main() {
    cout << "------------------------Program "
            "starts------------------------"
         << l << l;
    LinkedList list;
    list.operations();
    return 0;
}
